src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,182 @@
+                IEventReferenceOperation eOperation => eOperation.Member,
+                _ => null,
+            };
+
+        private void TryGetOrCachePreviewFeaturesAttributesOnSymbol(IOperation operation, ISymbol symbol, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols, bool checkParents)
+        {
+            if (!requiresPreviewFeaturesSymbols.TryGetValue(symbol, out IOperation? _))
+            {
+                if (checkParents)
+                {
+                    var containingSymbol = symbol.ContainingSymbol;
+                    // Namespaces do not have attributes
+                    while (containingSymbol is INamespaceSymbol)
+                    {
+                        containingSymbol = containingSymbol.ContainingSymbol;
+                    }
+
+                    if (containingSymbol != null)
+                    {
+                        TryGetOrCachePreviewFeaturesAttributesOnSymbol(operation, containingSymbol, requiresPreviewFeaturesSymbols, checkParents);
+                    }
+                }
+                if (symbol is ITypeSymbol typeSymbol)
+                {
+                    ImmutableArray<INamedTypeSymbol> interfaces = typeSymbol.AllInterfaces;
+                    foreach (INamedTypeSymbol? anInterface in interfaces)
+                    {
+                        TryGetOrCachePreviewFeaturesAttributesOnSymbol(operation, anInterface, requiresPreviewFeaturesSymbols, checkParents);
+                    }
+
+                    INamedTypeSymbol? baseType = typeSymbol.BaseType;
+                    if (baseType != null)
+                    {
+                        TryGetOrCachePreviewFeaturesAttributesOnSymbol(operation, baseType, requiresPreviewFeaturesSymbols, checkParents);
+                    }
+                }
+                if (symbol.IsOverride)
+                {
+                    ISymbol? overriddenMember = symbol.GetOverriddenMember();
+                    if (overriddenMember != null)
+                    {
+                        TryGetOrCachePreviewFeaturesAttributesOnSymbol(operation, overriddenMember, requiresPreviewFeaturesSymbols, checkParents);
+                    }
+                }
+                GetOrAddAttributes(symbol, operation, requiresPreviewFeaturesSymbols);
+            }
+        }
+
+        private void GetOrAddAttributes(ISymbol symbol, IOperation operation, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols)
+        {
+            IOperation? value = null;
+            ImmutableArray<AttributeData> attributes = symbol.GetAttributes();
+            foreach (AttributeData attribute in attributes)
+            {
+                string attributeName = attribute.AttributeClass.Name;
+                if (attributeName == RequiresPreviewFeaturesAttribute)
+                {
+                    value = operation;
+                    thisAssemblyUsesPreviewFeatures = true;
```

Comment:
Yep, this field is the reason for all race conditions that you might be seeing

Reply:
Seems its better to make this method return boolean instead which will be passed all the way to the `CheckOperationAttributes` to signal that it found an API having the attribute